File: app/src/main/java/com/formichelli/dfreminder/DFReminderNotificationListener.java
package com.formichelli.dfreminder;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.service.notification.NotificationListenerService;
import android.service.notification.StatusBarNotification;
import android.util.Log;

import java.util.Collection;
import java.util.HashSet;
import java.util.Set;

/**
 * Listen for notifications
 * <p>
 * Created by daniele on 29/10/15.
 */
public class DFReminderNotificationListener extends NotificationListenerService {
    private Collection<String> notifications;
    private RemindTimer remindTimer;
    private SharedPreferences sharedPreferences;
    private String isEnabledPreferenceString;
    private Set<String> packageWhiteList;

    @Override
    public void onCreate() {
        final Context context = getApplicationContext();
        notifications = new HashSet<>();
        remindTimer = new RemindTimer(getApplicationContext(), notifications);
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
        isEnabledPreferenceString = context.getString(R.string.pref_key_enable_dfreminder);
        packageWhiteList = new HashSet<>();
        packageWhiteList.add("android");
        packageWhiteList.add("com.android.providers.downloads");
        packageWhiteList.add("com.android.vending");
    }

    @Override
    public void onNotificationPosted(StatusBarNotification sbn) {
        synchronized (remindTimer) {
            if (shouldBeIgnored(sbn, true))
                return;

            notifications.add(sbn.getPackageName());
            remindTimer.restartReminderTimer();
        }
    }

    @Override
    public void onNotificationRemoved(StatusBarNotification sbn) {
        synchronized (remindTimer) {
            if (shouldBeIgnored(sbn, false))
                return;

            notifications.remove(sbn.getPackageName());
            if (notifications.isEmpty())
                remindTimer.stopReminderTimerIfRunning();
        }
    }

    private boolean shouldBeIgnored(StatusBarNotification sbn, boolean posted) {
        final String prefix = posted ? "New" : "Removed";
        final String packageName = sbn.getPackageName();

        if (!sharedPreferences.getBoolean(isEnabledPreferenceString, false)) {
            Log.d(DFReminderMainActivity.TAG, prefix + " notification from " + packageName + ": DF reminder disabled");
            return true;
        }

        if (packageWhiteList.contains(packageName)) {
            Log.d(DFReminderMainActivity.TAG, prefix + " notification from " + packageName + " ignored");
            return true;
        }

        if (!sbn.isClearable()) {
            Log.d(DFReminderMainActivity.TAG, prefix + " not clearable notification from " + packageName + " ignored");
            return true;
        }

        Log.d(DFReminderMainActivity.TAG, prefix + " notification from " + packageName + ", total: " + notifications.size());

        return false;
    }
}

File: app/src/main/java/com/formichelli/dfreminder/DFReminderNotificationListener.kt
package com.formichelli.dfreminder

import android.content.SharedPreferences
import android.preference.PreferenceManager
import android.service.notification.NotificationListenerService
import android.service.notification.StatusBarNotification
import android.util.Log
import java.util.*

/**
 * Listen for notifications
 *
 *
 * Created by daniele on 29/10/15.
 */
class DFReminderNotificationListener : NotificationListenerService() {
    private var notifications = HashSet<String>()
    private var remindTimer: RemindTimer? = null
    private var sharedPreferences: SharedPreferences? = null
    private var isEnabledPreferenceString: String? = null
    private var packageWhiteList = setOf("android", "com.android.providers.downloads", "com.android.vending", "com.android.systemui")

    override fun onCreate() {
        remindTimer = RemindTimer(applicationContext, notifications)
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(applicationContext)
        isEnabledPreferenceString = applicationContext.getString(R.string.pref_key_enable_dfreminder)
    }

    override fun onNotificationPosted(sbn: StatusBarNotification) {
        synchronized(remindTimer!!) {
            if (shouldBeIgnored(sbn, true))
                return

            notifications.add(sbn.packageName)
            remindTimer!!.restartReminderTimer()
        }
    }

    override fun onNotificationRemoved(sbn: StatusBarNotification) {
        synchronized(remindTimer!!) {
            if (shouldBeIgnored(sbn, false))
                return

            notifications.remove(sbn.packageName)
            if (notifications.isEmpty())
                remindTimer!!.stopReminderTimerIfRunning()
        }
    }

    private fun shouldBeIgnored(sbn: StatusBarNotification, posted: Boolean): Boolean {
        val prefix = if (posted) "New" else "Removed"
        val packageName = sbn.packageName

        if (!sharedPreferences!!.getBoolean(isEnabledPreferenceString, false)) {
            Log.d(DFReminderMainActivity.TAG, "$prefix notification from $packageName: DF reminder disabled")
            return true
        }

        if (packageWhiteList.contains(packageName)) {
            Log.d(DFReminderMainActivity.TAG, "$prefix notification from $packageName ignored")
            return true
        }

        if (!sbn.isClearable) {
            Log.d(DFReminderMainActivity.TAG, "$prefix not clearable notification from $packageName ignored")
            return true
        }

        Log.d(DFReminderMainActivity.TAG, "$prefix notification from $packageName, total: " + notifications.size)

        return false
    }
}
